paddle/fluid/framework/parallel_executor.cc

```diff
@@ -0,0 +1,415 @@
+          mems;
+      mems.emplace_back(const_cast<void *>(main_tensor.data<void>()),
+                        &member_->GetNCCLCtx(member_->main_place_));
+
+      for (auto &pair : member_->local_scopes_) {
+        if (pair.first == member_->main_place_) {
+          continue;
+        }
+
+        auto local_scope = pair.second;
+        auto *t = local_scope->Var(var_desc->Name())->GetMutable<LoDTensor>();
+        t->Resize(dims);
+        mems.emplace_back(t->mutable_data(pair.first, main_tensor.type()),
+                          &member_->GetNCCLCtx(member_->main_place_));
+      }
+
+      // TODO(yy): Invoke ncclBCast here. mems, numel, data_type. The mems[0]
+      // is the src, rests are dests.
+
+      (void)(data_type);
+      (void)(numel);
+    }
+  }
+#else
+  PADDLE_THROW("Not compiled with CUDA");
+#endif
+}
+
+void ParallelExecutor::BuildNCCLCommunicator() const {
+#ifdef PADDLE_WITH_CUDA
+  for (auto &place_pair : member_->local_scopes_) {
+    auto place = place_pair.first;
+    int dev_id = boost::get<platform::CUDAPlace>(place).device;
+
+    member_->communication_streams_.emplace(
+        dev_id, ParallelExecutorPrivate::NCCLContext(dev_id));
+  }
+
+  ParallelExecutorPrivate::NCCLContext::InitNCCLContext(
+      member_->communication_streams_);
+#endif
+}
+
+std::vector<LoDTensor> ParallelExecutor::Run(
+    const std::vector<std::string> &fetch_tensors) {
+  // Version --> VarHandle
+
+  std::unordered_map<VarHandle *, bool> pending_vars;
+  std::unordered_map<OpHandle *, size_t> pending_ops;
+
+  for (auto &place_pair : member_->vars_) {
+    for (auto &name_pair : place_pair.second) {
+      for (auto &version_pair : name_pair.second) {
+        pending_vars[&version_pair.second] =
+            version_pair.second.generated_op_ == nullptr;
+      }
+    }
+  }
+
+  for (auto &op : member_->ops_) {
+    pending_ops.insert({op.get(), op->inputs_.size()});
+  }
+
+  while (!pending_ops.empty()) {
+    VarHandle *ready_var = nullptr;
+    for (auto &pair : pending_vars) {
+      if (pair.second) {
+        ready_var = pair.first;
+      }
+    }
+
+    if (ready_var == nullptr) {
+      member_->pool_.Wait();  // Wait thread pool;
+      continue;
+    }
+
+    pending_vars.erase(ready_var);
+
+    std::vector<OpHandle *> to_run;
+
+    for (auto *op : ready_var->pending_ops_) {
+      auto &deps = pending_ops[op];
+      --deps;
+      if (deps == 0) {
+        to_run.emplace_back(op);
+      }
+    }
+
+    for (auto *op : to_run) {
+      pending_ops.erase(op);
+
+      std::vector<bool *> ready_buffer;
+      for (auto *var : op->outputs_) {
+        ready_buffer.emplace_back(&pending_vars[var]);
+      }
+
+      auto op_run = [ready_buffer, op] {
+        // TODO(yy) Check Previous Op has same dev ctx.
+        LOG(INFO) << "Run " << op->DebugString();
+        for (auto *ready : ready_buffer) {
+          *ready = true;
+        }
+      };
+
+      member_->pool_.Run(op_run);
```

Comment:
Actually, we should add a callback after we push operator run job to memory pool. In this callback, we change pending_var state.

Reply:
The pending_var state has been changed during L404-L406